### PR TITLE
Swift: Query for regular expression injection

### PR DESCRIPTION
**Pull request by geoffw0**
Adds a query for regular expression injection, that is, constructing a regular expression from unsanitized user input.  This is fairly straightforward as the vulnerability is a simple injection issue and we already have a regular expressions library that identifies points where regular expressions are evaluated.  It turns out we also want to identify points where regular expressions are constructed in advance of evaluation, so as to get clean and reliable alerts, so we do that as well.

There are a handful of test results that are false positives or not detected.  We have follow-up issues to address all of these, and in the case of the false positives I do not believe they will be noisy in real world projects before we address them.

TODO:
- [x] team review
- [x] doc review
- [x] DCA run

### CODE REVIEW

**Comment by geoffw0** [comment]
DCA run LGTM.

File: swift/ql/src/queries/Security/CWE-730/RegexInjectionGood.swift [hunk 1]
```diff
@@ -0,0 +1,13 @@
+func processRemoteInput(remoteInput: String) {
+  ...
+
+  # GOOD: Regular expression is not derived from user input
```
Comment (rdmarsh2) [inline swift/ql/src/queries/Security/CWE-730/RegexInjectionGood.swift:4]: Shouldn't these be `//` comments?

Comment (geoffw0) [reply swift/ql/src/queries/Security/CWE-730/RegexInjectionGood.swift]: Good spot.  Fixed.

File: swift/ql/src/queries/Security/CWE-730/RegexInjection.qhelp [hunk 2]
```diff
@@ -0,0 +1,50 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+
+<overview>
+<p>
+Constructing a regular expression with unsanitized user input is dangerous,
+since a malicious user may be able to modify the meaning of the expression. They
+may be able to cause unexpected program behaviour, or perform a Denial of Service
```
Comment (mattpollard) [inline swift/ql/src/queries/Security/CWE-730/RegexInjection.qhelp:10]: Not sure if we typically capitalize the 'D' and 'S' here.

```suggestion
may be able to cause unexpected program behaviour, or perform a denial-of-service
```

File: swift/ql/src/queries/Security/CWE-730/RegexInjection.qhelp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+attack. For example they may provide a regular expression fragment that takes
```
Comment (mattpollard) [inline swift/ql/src/queries/Security/CWE-730/RegexInjection.qhelp:11]:
```suggestion
attack. For example, they may provide a regular expression fragment that takes
```

**Review by mattpollard — DISMISSED** [review]

File: swift/ql/src/queries/Security/CWE-730/RegexInjection.qhelp (hunk 2 above)
Comment (geoffw0) [reply swift/ql/src/queries/Security/CWE-730/RegexInjection.qhelp:10]: In the CodeQL repo we have both "Denial of Service" and "denial-of-service" plenty of times, with little consistency.  I agree the capitalization seems a little excessive.

**Comment by geoffw0** [comment]
Thanks for the review @mattpollard.  I've accepted both of your suggestions.

@rdmarsh2 I did your recommended change as well, are you happy to approve this now?

**Review by rdmarsh2 — APPROVED** [review]

**Pull request merged**